File: 백준/Silver/2839. 설탕 배달/설탕 배달.cc
#include <iostream>
using namespace std;

int sugar[5001] = { 0 };

int howManySugarBag(int n) {
    if (n == 3 || n == 5) {
        sugar[n] = 1;
    }
    else if (n == 4 || n == 7) {
        sugar[n] = -1;
    }
    else if (n >= 6) {
        if (n % 5 == 0){
            sugar[n] = (n / 5);
        }
        else if (sugar[n - 3] != -1){
            sugar[n] = howManySugarBag(n - 3) + 1;
        }
    }

    return sugar[n];
}

int main() {
    int n; cin >> n;
    cout << howManySugarBag(n) << endl;

    return 0;
}

File: 백준/Silver/2839. 설탕 배달/설탕 배달.java
import java.util.Scanner;

public class Main {
	static int[] sugar;

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		sc.close();
		sugar = new int[n + 1];
		
		System.out.println(howManySugarBag(n));
	}

	static int howManySugarBag(int n) {
		if(n == 3 || n == 5) 
			sugar[n] = 1;
		
		if(n == 4 || n == 7)
			sugar[n] = -1;
		
		else {
			if(n % 5 == 0)
				sugar[n] = (n / 5);

			else if(sugar[n - 3] != -1)
				sugar[n] = howManySugarBag(n - 3) + 1;
		}
		
		return sugar[n];
	}
}